File: backend/app/models/stamp.py
import uuid
from datetime import datetime
from sqlalchemy import String, Boolean, DateTime, Text, Integer, ForeignKey, JSON
from sqlalchemy.orm import Mapped, mapped_column
from app.database import Base


class Stamp(Base):
    __tablename__ = "stamps"

    id: Mapped[str] = mapped_column(
        String(36), primary_key=True, default=lambda: str(uuid.uuid4())
    )
    user_id: Mapped[str] = mapped_column(
        String(36), ForeignKey("users.id"), nullable=False, index=True
    )
    original_image_url: Mapped[str | None] = mapped_column(Text, nullable=True)
    stamp_image_url: Mapped[str | None] = mapped_column(Text, nullable=True)
    thumbnail_url: Mapped[str | None] = mapped_column(Text, nullable=True)
    title: Mapped[str | None] = mapped_column(String(200), nullable=True)
    description: Mapped[str | None] = mapped_column(Text, nullable=True)
    stamp_style: Mapped[dict | None] = mapped_column(JSON, nullable=True)
    tags: Mapped[list | None] = mapped_column(JSON, nullable=True)
    is_public: Mapped[bool] = mapped_column(Boolean, default=True)
    likes_count: Mapped[int] = mapped_column(Integer, default=0)
    comments_count: Mapped[int] = mapped_column(Integer, default=0)
    views_count: Mapped[int] = mapped_column(Integer, default=0)
    created_at: Mapped[datetime] = mapped_column(DateTime, default=datetime.utcnow)
    updated_at: Mapped[datetime] = mapped_column(
        DateTime, default=datetime.utcnow, onupdate=datetime.utcnow
    )

File: lib/features/stamp/domain/entities/stamp.dart
class Stamp {
  final String id;
  final String userId;
  final String? originalImageUrl;
  final String? stampImageUrl;
  final String? thumbnailUrl;
  final String? title;
  final String? description;
  final Map<String, dynamic>? stampStyle;
  final List<String>? tags;
  final bool isPublic;
  final int likesCount;
  final int commentsCount;
  final int viewsCount;
  final DateTime createdAt;
  final DateTime? updatedAt;

  const Stamp({
    required this.id,
    required this.userId,
    this.originalImageUrl,
    this.stampImageUrl,
    this.thumbnailUrl,
    this.title,
    this.description,
    this.stampStyle,
    this.tags,
    this.isPublic = true,
    this.likesCount = 0,
    this.commentsCount = 0,
    this.viewsCount = 0,
    required this.createdAt,
    this.updatedAt,
  });

  factory Stamp.fromJson(Map<String, dynamic> json) {
    return Stamp(
      id: json['id'],
      userId: json['user_id'],
      originalImageUrl: json['original_image_url'],
      stampImageUrl: json['stamp_image_url'],
      thumbnailUrl: json['thumbnail_url'],
      title: json['title'],
      description: json['description'],
      stampStyle: json['stamp_style'] != null
          ? Map<String, dynamic>.from(json['stamp_style'])
          : null,
      tags: json['tags'] != null ? List<String>.from(json['tags']) : null,
      isPublic: json['is_public'] ?? true,
      likesCount: json['likes_count'] ?? 0,
      commentsCount: json['comments_count'] ?? 0,
      viewsCount: json['views_count'] ?? 0,
      createdAt: DateTime.parse(json['created_at']),
      updatedAt: json['updated_at'] != null
          ? DateTime.parse(json['updated_at'])
          : null,
    );
  }
}
